File: Emu/cpu.cpp
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <fstream>

#include <SDL.h>
#include "cpu.h"
#include "editor.h"
#include "timing.h"
#include "graphics.h"


#ifdef _WIN32
#include <Windows.h>
#endif


namespace Cpu
{
    int64_t _clock = -2;
    uint8_t _IN = 0xFF, _XOUT = 0x00;
    uint8_t _ROM[ROM_SIZE][2], _RAM[RAM_SIZE];


    int64_t getClock(void) {return _clock;}
    uint8_t getIN(void) {return _IN;}
    uint8_t getXOUT(void) {return _XOUT;}
    uint8_t getROM(uint16_t address, int page) {return _ROM[address & (ROM_SIZE-1)][page & 0x01];}
    uint8_t getRAM(uint16_t address) {return _RAM[address & (RAM_SIZE-1)];}

    void setClock(int64_t clock) {_clock = clock;}
    void setIN(uint8_t in) {_IN = in;}
    void setXOUT(uint8_t xout) {_XOUT = xout;}
    void setRAM(uint16_t address, uint8_t data)
    {
        // Constant "0" and "1" are stored here
        if(address == 0x0000) return;
        if(address == 0x0080) return;

        _RAM[address & (RAM_SIZE-1)] = data;
    }

    void setROM(uint16_t base, uint16_t address, uint8_t data)
    {
        uint16_t offset = (address - base) / 2;
        _ROM[base + offset][address & 0x01] = data;
    }

    void garble(uint8_t* mem, int len)
    {
        for(int i=0; i<len; i++) mem[i] = rand();
    }

    void initialise(Cpu::State& S)
    {
#ifdef _WIN32
        CONSOLE_SCREEN_BUFFER_INFO csbi;

        if(!AllocConsole()) return;

        if (GetConsoleScreenBufferInfo(GetStdHandle(STD_OUTPUT_HANDLE), &csbi))
        {
            csbi.dwSize.Y = 1000;
            SetConsoleScreenBufferSize(GetStdHandle(STD_OUTPUT_HANDLE), csbi.dwSize);
        }

        if(!freopen("CONIN$", "w", stdin)) return;
        if(!freopen("CONOUT$", "w", stderr)) return;
        if (!freopen("CONOUT$", "w", stdout)) return;
        setbuf(stdout, NULL);
#endif    

        // Memory
        srand(unsigned int(time(NULL))); // Initialize with randomized data
        garble((uint8_t*)_ROM, sizeof _ROM);
        garble(_RAM, sizeof _RAM);
        garble((uint8_t*)&S, sizeof S);

        // Open ROM file
        std::ifstream romfile("theloop.2.rom", std::ios::binary | std::ios::in);
        if(!romfile.is_open())
        {
            fprintf(stderr, "Cpu::initialise() : failed to open ROM file\n");
            exit(EXIT_FAILURE);
        }

        // Read ROM file
        romfile.read((char *)_ROM, sizeof(_ROM));
        if(romfile.bad() || romfile.fail())
        {
            fprintf(stderr, "Cpu::initialise() : failed to read ROM file\n");
            exit(EXIT_FAILURE);
        }

        // SDL initialisation
        if(SDL_Init(SDL_INIT_VIDEO | SDL_INIT_AUDIO | SDL_INIT_EVENTS) < 0)
        {
            fprintf(stderr, "Cpu::initialise() : failed to initialise SDL\n");
            exit(EXIT_FAILURE);
        }
    }

    State cycle(const State& S)
    {
        State T = S; // New state is old state unless something changes
    
        T._IR = _ROM[S._PC][0]; // Instruction Fetch
        T._D  = _ROM[S._PC][1];
    
        int ins = S._IR >> 5;       // Instruction
        int mod = (S._IR >> 2) & 7; // Addressing mode (or condition)
        int bus = S._IR&3;          // Busmode
        int W = (ins == 6);        // Write instruction?
        int J = (ins == 7);        // Jump instruction?
    
        uint8_t lo=S._D, hi=0, *to=NULL; // Mode Decoder
        int incX=0;
        if(!J)
        {
            switch (mod)
            {
                #define E(p) (W?0:p) // Disable _AC and _OUT loading during _RAM write
                case 0: to=E(&T._AC);                            break;
                case 1: to=E(&T._AC); lo=S._X;                   break;
                case 2: to=E(&T._AC);         hi=S._Y;           break;
                case 3: to=E(&T._AC); lo=S._X; hi=S._Y;          break;
                case 4: to=  &T._X;                              break;
                case 5: to=  &T._Y;                              break;
                case 6: to=E(&T._OUT);                           break;
                case 7: to=E(&T._OUT); lo=S._X; hi=S._Y; incX=1; break;
            }
        }

        uint16_t addr = (hi << 8) | lo;
        int B = S._undef; // Data Bus
        switch(bus)
        {
            case 0: B=S._D;                              break;
            case 1: if (!W) B = _RAM[addr&(RAM_SIZE-1)]; break;
            case 2: B=S._AC;                             break;
            case 3: B=_IN;                               break;
        }

        if(W) _RAM[addr&(RAM_SIZE-1)] = B; // Random Access Memory

        uint8_t ALU; // Arithmetic and Logic Unit
        switch(ins)
        {
            case 0: ALU =        B; break; // LD
            case 1: ALU = S._AC & B; break; // ANDA
            case 2: ALU = S._AC | B; break; // ORA
            case 3: ALU = S._AC ^ B; break; // XORA
            case 4: ALU = S._AC + B; break; // ADDA
            case 5: ALU = S._AC - B; break; // SUBA
            case 6: ALU = S._AC;     break; // ST
            case 7: ALU = -S._AC;    break; // Bcc/JMP
        }

        if(to) *to = ALU; // Load value into register
        if(incX) T._X = S._X + 1; // Increment _X

        T._PC = S._PC + 1; // Next instruction
        if(J)
        {
            if(mod != 0) // Conditional branch within page
            {
                int cond = (S._AC>>7) + 2*(S._AC==0);
                if (mod & (1 << cond)) // 74153
                T._PC = (S._PC & 0xff00) | B;
            }
            else
            {
                T._PC = (S._Y << 8) | B; // Unconditional far jump
            }
        }

        return T;
    }

    void reset(bool coldBoot)
    {
        // Cold boot
        if(coldBoot)
        {
            setRAM(BOOT_COUNT, 0x00);
            setRAM(BOOT_CHECK, 0xA6); // TODO: don't hardcode the checksum, calculate it properly
        }

        Graphics::resetVTable();
        Editor::setSingleStepWatchAddress(VIDEO_Y_ADDRESS);
        setClock(CLOCK_RESET);
    }
}

File: Contrib/tbraun-de/pytron/cpu.py
from random import randint
import sys


class CpuState:
    def __init__(self):
        self.pc = randint(0, 255)
        self.ir = randint(0, 255)
        self.d = randint(0, 255)
        self.ac = randint(0, 255)
        self.x = randint(0, 255)
        self.y = randint(0, 255)
        self.out = randint(0, 255)
        self.undef = randint(0, 255)

    def copyTo(self, destination):
        destination.pc = self.pc
        destination.ir = self.ir
        destination.d = self.d
        destination.ac = self.ac
        destination.x = self.x
        destination.y = self.y
        destination.out = self.out
        destination.undef = self.undef


class Cpu():
    def __init__(self):
        self.rom = [[randint(0, 255) for x in range(2)] for y in range(1 << 16)]
        self.ram = [randint(0, 255) for x in range(1 << 15)]
        self.IN = 0xff
        self.tempState = CpuState()

    def cpuCycle(self, state: CpuState):
        state.copyTo(self.tempState)

        self.tempState.ir = self.rom[state.pc][0]
        self.tempState.d = self.rom[state.pc][1]

        instruction = state.ir >> 5
        mod = (state.ir >> 2) & 7
        bus = state.ir & 3
        w = (instruction == 6)
        j = (instruction == 7)
        lo = state.d
        hi = 0
        to = None

        inc_x = False

        if not j:
            if mod == 0:
                to = "AC" if w is False else None
            elif mod == 1:
                to = "AC" if w is False else None
                lo = state.x
            elif mod == 2:
                to = "AC" if w is False else None
                hi = state.y
            elif mod == 3:
                to = "AC" if w is False else None
                lo = state.x
                hi = state.y
            elif mod == 4:
                to = "X"
            elif mod == 5:
                to = "Y"
            elif mod == 6:
                to = "OUT" if w is False else None
            elif mod == 7:
                to = "OUT" if w is False else None
                lo = state.x
                hi = state.y
                inc_x = True
        addr = (hi << 8) | lo

        b = state.undef
        if bus == 0:
            b = state.d
        elif bus == 1:
            if not w:
                b = self.ram[addr & 0x7fff]
        elif bus == 2:
            b = state.ac
        elif bus == 3:
            b = self.IN

        if w:
            self.ram[addr & 0x7fff] = b

        if instruction == 0:    # LD
            ALU = b
        elif instruction == 1:  # ANDA
            ALU = state.ac & b
        elif instruction == 2:  # ORA
            ALU = state.ac | b
        elif instruction == 3:  # XORA
            ALU = state.ac ^ b
        elif instruction == 4:  # ADDA
            ALU = (state.ac + b) & 0xFF
        elif instruction == 5:  # SUBA
            ALU = (state.ac - b) & 0xFF
        elif instruction == 6:  # ST
            ALU = state.ac
        elif instruction == 7:  # Bcc/JMP
            ALU = (-state.ac) & 0xFF

        # Load value into register
        if to == "AC":
            self.tempState.ac = ALU
        elif to == "X":
            self.tempState.x = ALU
        elif to == "Y":
            self.tempState.y = ALU
        elif to == "OUT":
            self.tempState.out = ALU

        if inc_x:
            self.tempState.x = (state.x + 1) & 0xFF

        self.tempState.pc = (state.pc + 1) & 0xFFFF
        if j:
            if mod != 0:
                cond = (state.ac >> 7) + 2 * (state.ac == 0)
                if (mod & (1 << cond)) != 0:    # 74153
                    self.tempState.pc = (state.pc & 0xff00) | b
            else:
                self.tempState.pc = (state.y << 8) | b   # Unconditional far jump
        return self.tempState


state = CpuState()
new_state = CpuState()
cpu = Cpu()
with open("../../../ROMv3.rom", "rb") as f:
    byte = f.read(1)
    rom_address = 0
    while byte != b"":
        cpu.rom[rom_address][0] = ord(byte)
        byte = f.read(1)
        cpu.rom[rom_address][1] = ord(byte)
        byte = f.read(1)
        rom_address = rom_address + 1

vgaX = 0
vgaY = 0

t = -2
while True:
    if t < 0:
        state.pc = 0  # MCP100 Power-On Reset

    #sys.stdout.write("PC %04x: %02x%02x IR:%02x D:%02x AC:%02x X:%02x Y:%02x OUT:%02x UNDEF:%02x\n" %
    #                 (state.pc, cpu.rom[state.pc][0], cpu.rom[state.pc][1], state.ir, state.d, state.ac, state.x,
    #                  state.y, state.out, state.undef))

    cpu.cpuCycle(state).copyTo(new_state)  # Update CPU

    hSync = (new_state.out & 0x40) - (state.out & 0x40)  # "VGA monitor" (use simple stdout)
    vSync = (new_state.out & 0x80) - (state.out & 0x80)
    if vSync < 0:
        vgaY = -36
    vgaX = vgaX + 1
    if vgaX <= 200:
        if hSync != 0:
            sys.stdout.write('|')  # Visual indicator of hSync
        elif vgaX == 200:
            sys.stdout.write('>')  # too many pixels
        elif (~state.out) & 0x80 > 0:
            sys.stdout.write('^')  # visualize vBlank pulse
        else:
            sys.stdout.write(chr(32 + (state.out & 63)))  # plot pixel
    if hSync > 0:
        if vgaX != 200:
            sys.stdout.write('~')  # mark horizontal cycle errors
        sys.stdout.write("line %-3d xout %02x t %0.3f\n" % (vgaY, new_state.ac, t / 6250000))
        vgaX = 0
        vgaY = vgaY + 1
        new_state.undef = randint(0, 255)

    new_state.copyTo(state)

    t = t + 1

